### DISCUSSION TITLE
I want to setup a livestream using the RTMP and NGINX server for a DJI drone.

### DISCUSSION THREAD

**Discussion by ShabanWaris1234** [ultralytics/ultralytics#12839] [Q&A] [answered]
I want to set up a live stream using the RTMP and NGINX server for a DJI drone. I have already trained a custom yolov8 model. However, I am having a problem setting up a live stream with the RTMP and NGINX server for my drone. Are there any tutorials or codes that can guide me in setting up a real-time detection using a drone? I also need to make a web app for my project I plan to use flask. Any tutorials would help.

**Comment by glenn-jocher** [comment, marked as answer]
Certainly! Setting up a live stream with RTMP and NGINX for your DJI drone and integrating it with your custom YOLOv8 model for real-time detection sounds like an exciting project! Here’s a brief guide to get you started:

1. **RTMP Server Setup with NGINX:**
   - Install NGINX and the RTMP module. You can find a good tutorial on how to do this [here](https://www.nginx.com/blog/video-streaming-for-remote-learning-with-nginx/).
   - Configure NGINX to receive the RTMP stream from your drone. You'll need to modify the `nginx.conf` file to include the RTMP module settings.

2. **Integrate YOLOv8 for Real-Time Detection:**
   - Use the Python `cv2` library to capture video frames from the RTMP stream.
   - Load your trained YOLOv8 model and pass the video frames to it for detection.
   ```python
   import cv2
   from ultralytics import YOLO

   # Load model
   model = YOLO('path/to/your/custom_model.pt')

   # Capture RTMP stream
   cap = cv2.VideoCapture('rtmp://localhost/live/stream_name')
   while cap.isOpened():
       ret, frame = cap.read()
       if ret:
           # Make predictions
           results = model(frame)
           results.show()  # Display the frame with detections
       if cv2.waitKey(1) & 0xFF == ord('q'):
           break
   cap.release()
   cv2.destroyAllWindows()
   ```

3. **Developing a Web App with Flask:**
   - Flask is a great choice for setting up a web server. You can start by creating a basic Flask app that displays the video stream.
   - Integrate the Flask app with the detection model to display the processed frames.
   - Check out the [Flask documentation](https://flask.palletsprojects.com/en/2.0.x/) for basic setup and [this tutorial](https://blog.miguelgrinberg.com/post/video-streaming-with-flask) for streaming video through Flask.

These steps should help you set up your project. For specific issues or errors, feel free to ask for more detailed help! Good luck with your real-time detection system! 🚀

**Comment by mordka** [comment]
@ShabanWaris1234 just a note, you won't get realtime detection, because RTMP has got few seconds latency embedded in protocol design. I've built realtime AI drone streaming pipeline with <1s AI latency and I'm open to share the details. If you are interested in this reach out to me on mordka@proton.me

**Reply by ccamacho** [reply]
you could share it here

**Reply by glenn-jocher** [reply]
Hello @ccamacho,

Thank you for your interest! For setting up a live stream with RTMP and NGINX for your DJI drone, you can follow these general steps:

1. **Set up NGINX with RTMP Module**: Configure NGINX on your server to handle RTMP streams. This involves installing the NGINX server with the RTMP module and configuring it to accept live streams.

2. **Stream from Drone to NGINX**: Use your DJI drone's software to send the video feed to your NGINX server's RTMP URL. This typically involves setting the RTMP stream URL in the drone's broadcasting settings.

3. **Process Stream with YOLOv8**: Capture the RTMP stream in your application using a library like OpenCV, and then process the frames with your custom-trained YOLOv8 model for real-time object detection.

4. **Web App Integration**: For integrating with a Flask web app, you can set up Flask to serve the processed video stream to a frontend. You might use WebSocket or another method to push the video frames to the client browser in near real-time.

For detailed code examples and step-by-step guides, you might want to check out resources specific to each component (NGINX RTMP setup, OpenCV video capture, Flask streaming). Unfortunately, I don't have a specific tutorial to share here, but these keywords should help you find relevant resources.

If you encounter specific issues or have questions about integrating these components, feel free to ask here!

Best of luck with your project! 🚀

— The Ultralytics Team

**Reply by ShabanWaris1234** [reply]
Thanks for the help @glenn-jocher.
Was able to setup the stream.
Any tips on how to lower the latency. The latency normally is around 3-4seconda delay. But when the model is inferring the latency is 40 seconds+

**Reply by glenn-jocher** [reply]
Hi @ShabanWaris1234,

Great to hear that you were able to set up the stream! To address the latency issue, here are a few tips that might help:

1. **Optimize Model Inference**:
   - Ensure you are using the latest versions of `torch` and `ultralytics`. Upgrading can sometimes bring performance improvements.
   - Consider using a smaller model variant (e.g., YOLOv8n instead of YOLOv8x) to reduce inference time.
   - Use FP16 or INT8 quantization to speed up inference. You can export your model with these optimizations using:
     ```python
     model.export(format='onnx', half=True)  # for FP16
     model.export(format='onnx', int8=True)  # for INT8
     ```

2. **Stream Configuration**:
   - Reduce the resolution and frame rate of your stream to lower the processing load.
   - Use a more efficient video codec (e.g., H.264) and adjust the bitrate settings.

3. **Hardware Acceleration**:
   - Utilize GPU acceleration if available. Ensure your inference code is running on the GPU:
     ```python
     device = 'cuda' if torch.cuda.is_available() else 'cpu'
     model.to(device)
     ```

4. **Buffer Management**:
   - Adjust the buffer settings in your RTMP server configuration to minimize delay.

5. **Network Optimization**:
   - Ensure a stable and high-speed network connection to reduce transmission delays.

If you continue to experience issues, please provide a minimum reproducible code example so we can investigate further. You can find more details on creating one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Hope this helps! 😊

**Reply by kylegraupe** [reply]
Does this solution work for all DJI Drones? I'd like to use this solution for a DJI Mini 4, if possible. @glenn-jocher

**Reply by glenn-jocher** [reply]
The solution should work with any DJI drone that supports RTMP streaming, including the DJI Mini 4. Ensure your drone's firmware supports this feature.

**Reply by kylegraupe** [reply]
Hi all,

It looks like I was in fact able to find a solution to this problem. You can check it out on [this github repo](https://github.com/kylegraupe/uas-rtcv). It took me some time to get it fully functioning, but I have posted the open-source code for everyone to see. If you would like to contribute, please do! There are application trailers on my website that show the application in action. If you see any issues, please add an issue to the board or fork the repo!

@glenn-jocher @ShabanWaris1234 @mordka

**Reply by glenn-jocher** [reply]
Thank you for sharing your solution and open-sourcing the code! This is a great contribution to the community. For any questions or feedback, users can engage directly on your GitHub repo. Keep up the great work!

**Reply by kylegraupe** [reply]
Thank you! 

Please leave a star or a follow if you find it helpful! It means a lot for the code to be discovered by others looking to contribute. @glenn-jocher

**Reply by kylegraupe** [reply]
If you'd like to learn more, please visit my website [here](https://www.graupe.io/portfolio/real-time-computer-vision-streamed-via-dji-drone) where you can see recordings from various test flights. If you have any specific questions regarding the use of the application, have any requests, or would like to work on a project together, my inbox is open!

Below are a few screenshots of a from test flights, where I have a segmentation mask next to the original frames in 720p (actually 704x1280 due to U-Net resizing necessity). 

![uas_rtcv_test_flight](https://github.com/user-attachments/assets/6b717654-75de-4418-957d-0a8097e2173c)

P.S. - I am currently looking for employment as a ML Engineer or Data Scientist. I am pursuing my second Master's degree, M.S. in Machine Learning, and already hold two degrees: M.S. in Modeling and Simulation Engineering and a Bachelor's in Aerospace and Ocean Engineering. Should this project impress, please let me know if there is a good fit!

@glenn-jocher @mordka @ShabanWaris1234 @ccamacho

**Reply by glenn-jocher** [reply]
Thank you for sharing your work and resources with the community! Unfortunately, we’re unable to assist with private collaboration or employment requests, but your project is impressive, and we encourage others to explore your repository and website for more details. Best of luck with your endeavors!

**Reply by abdoukhadre-searching** [reply]
> Hi all,
> 
> 
> 
> It looks like I was in fact able to find a solution to this problem. You can check it out on [this github repo](https://github.com/kylegraupe/uas-rtcv). It took me some time to get it fully functioning, but I have posted the open-source code for everyone to see. If you would like to contribute, please do! There are application trailers on my website that show the application in action. If you see any issues, please add an issue to the board or fork the repo!
> 
> 
> 
> @glenn-jocher @ShabanWaris1234 @mordka 

Hi I was just checking that repo for a projet similar where I have the same problem but the repos is no longer available 404

**Reply by engrjonel** [reply]
Hi! We’re currently working on a tree counting project using a DJI drone with live object detection (YOLO). Aside from the camera, do you have any tips or advice on what additional hardware we can mount on the drone to improve functionality or performance? Would love to hear your suggestions!
